### PR TITLE
toolchain: Add Espressif's ESP32 and ESP32S2 support

### PR DESCRIPTION
**Pull request by sylvioalves**

### CODE REVIEW

**Comment by sylvioalves** [comment]
@galak, not sure if the buildkite error is indeed an error. Would you check?

**Comment by sylvioalves** [comment]
@stephanosio @galak We analyzed all the patches in Espressif's gcc/gdb/newlib repository and concluded that we can have it as is (without patches). We will submit some patches upstream eventually, but there is no blocking patch at all to have it integrated in Zephyr.  
However, this won't work in Zephyr with latest branch, of course. I will submit a PR with a bunch of changes to enable the usage of this new toolchain. Can you suggest me a way to synchronize this?

**Comment by stephanosio** [comment]
> However, this won't work in Zephyr with latest branch, of course.

Do you mean the "ESP32 toolchain" built using the changes in this PR is not able to build working binaries for the ESP32 SoC? If so, this PR should be considered a draft until it can.

> Can you suggest me a way to synchronize this?

If you have ESP32-specific patches for the binutils, gcc and newlib, please create a pull request in the relevant Zephyr fork repositories and update this PR to reference them (you can set `*_DEVEL_REVISION` to `pull/(PR NUMBER)/head`).

**Comment by sylvioalves** [comment]
@stephanosio Yes, current main branch can't be build using the toolchain from this PR. I am working out a PR to enabled this (I have it all working locally already).
Regarding patches, I won't submit any patch at all as it works as is.

**Review by stephanosio — CHANGES_REQUESTED** [review]
Please specify the source (URL) and version of the `xtensa_espressif_esp32.tar.gz` and `xtensa_espressif_esp32s2.tar.gz` added here in the commit message.

**Comment by stephanosio** [comment]
Also noting that this needs to be tested against https://github.com/zephyrproject-rtos/zephyr/pull/43534/.

**Comment by sylvioalves** [comment]
> Please specify the source (URL) and version of the `xtensa_espressif_esp32.tar.gz` and `xtensa_espressif_esp32s2.tar.gz` added here in the commit message.

I added the github URL in the commit message. The .tar.gz file was manually created.

**Comment by stephanosio** [comment]
> I added the github URL in the commit message.

@sylvioalves Thanks. Please add the version/tag or commit (in `xtensa-overlay` repository) from which the tarball was created in the commit message as well. This will make identifying the current version of overlays we have in the sdk-ng repository much easier, in case we ever need to update these overlays in the future.

**Review by stephanosio — COMMENTED** [review]
Looks ok overall.

DNM until the build failures described in https://github.com/zephyrproject-rtos/zephyr/pull/43534#pullrequestreview-903701217 are resolved, in case they require Zephyr SDK-side changes.

**Comment by stephanosio** [comment]
Postponing this to 0.14.1, which will be a quick follow-up to 0.14.0 (multi-platform SDK) in a month or so.

File: configs/xtensa-espressif_esp32_zephyr-elf.config [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# CT_LOG_PROGRESS_BAR is not set
+CT_TARGET_CFLAGS="-mlongcalls"
+CT_ARCH_XTENSA=y
+CT_XTENSA_CUSTOM=y
+CT_OVERLAY_NAME="espressif_esp32"
+CT_OVERLAY_LOCATION="./overlays"
+CT_TARGET_VENDOR="espressif_esp32_zephyr"
+CT_BINUTILS_SRC_DEVEL=y
+CT_BINUTILS_DEVEL_URL="https://github.com/zephyrproject-rtos/binutils-gdb.git"
+CT_BINUTILS_DEVEL_REVISION="c7d30a54fc17c1e9d9b1c39c7623222442c63af8"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32_zephyr-elf.config:14]:
```suggestion
CT_BINUTILS_SRC_CUSTOM=y
CT_BINUTILS_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/binutils"
```

File: configs/xtensa-espressif_esp32_zephyr-elf.config [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+CT_NEWLIB_SRC_DEVEL=y
+CT_NEWLIB_DEVEL_URL="https://github.com/zephyrproject-rtos/newlib-cygwin.git"
+CT_NEWLIB_DEVEL_REVISION="2893dbb10f1fe3bf431a26f89a22e12a5020dc02"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32_zephyr-elf.config:17]:
```suggestion
CT_NEWLIB_SRC_CUSTOM=y
CT_NEWLIB_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/newlib"
```

File: configs/xtensa-espressif_esp32_zephyr-elf.config [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+CT_LIBC_NEWLIB_MULTITHREAD=y
+CT_LIBC_NEWLIB_RETARGETABLE_LOCKING=y
+# CT_LIBC_NEWLIB_WIDE_ORIENT is not set
+# CT_LIBC_NEWLIB_NANO_MALLOC is not set
+# CT_LIBC_NEWLIB_NANO_FORMATTED_IO is not set
+CT_LIBC_NEWLIB_EXTRA_SECTIONS=y
+CT_GCC_SRC_DEVEL=y
+CT_GCC_DEVEL_VCS_git=y
+CT_GCC_DEVEL_URL="https://github.com/zephyrproject-rtos/gcc.git"
+CT_GCC_DEVEL_REVISION="8c1f3dcc6eddf3392d77f56c9bed69e5d6446cdf"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32_zephyr-elf.config:36]:
```suggestion
CT_GCC_SRC_CUSTOM=y
CT_GCC_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gcc"
```

File: configs/xtensa-espressif_esp32_zephyr-elf.config [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+CT_CC_GCC_EXTRA_CONFIG_ARRAY="--with-gnu-ld --with-gnu-as --enable-initfini-array"
+CT_CC_LANG_CXX=y
+CT_DEBUG_GDB=y
+CT_GDB_SRC_DEVEL=y
+CT_GDB_DEVEL_URL="https://github.com/zephyrproject-rtos/binutils-gdb.git"
+CT_GDB_DEVEL_REVISION="76b05e962501d75d995b9f9dd13f837df36b7208"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32_zephyr-elf.config:42]:
```suggestion
CT_GDB_SRC_CUSTOM=y
CT_GDB_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gdb"
```

File: configs/xtensa-espressif_esp32_zephyr-elf.config [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+CT_GDB_CROSS_PYTHON_BINARY="python3.8"
+CT_GDB_CROSS_BUILD_NO_PYTHON=y
+CT_COMP_LIBS_NEWLIB_NANO=y
+CT_NEWLIB_NANO_SRC_DEVEL=y
+CT_NEWLIB_NANO_DEVEL_URL="https://github.com/zephyrproject-rtos/newlib-cygwin.git"
+CT_NEWLIB_NANO_DEVEL_REVISION="2893dbb10f1fe3bf431a26f89a22e12a5020dc02"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32_zephyr-elf.config:48]:
```suggestion
CT_NEWLIB_NANO_SRC_CUSTOM=y
CT_NEWLIB_NANO_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/newlib"
```

File: configs/xtensa-espressif_esp32s2_zephyr-elf.config [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+# CT_LOG_PROGRESS_BAR is not set
+CT_TARGET_CFLAGS="-mlongcalls"
+CT_ARCH_XTENSA=y
+CT_XTENSA_CUSTOM=y
+CT_OVERLAY_NAME="espressif_esp32s2"
+CT_OVERLAY_LOCATION="./overlays"
+CT_TARGET_VENDOR="espressif_esp32s2_zephyr"
+CT_BINUTILS_SRC_DEVEL=y
+CT_BINUTILS_DEVEL_URL="https://github.com/zephyrproject-rtos/binutils-gdb.git"
+CT_BINUTILS_DEVEL_REVISION="c7d30a54fc17c1e9d9b1c39c7623222442c63af8"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32s2_zephyr-elf.config:14]:
```suggestion
CT_BINUTILS_SRC_CUSTOM=y
CT_BINUTILS_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/binutils"
```

File: configs/xtensa-espressif_esp32s2_zephyr-elf.config [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+CT_NEWLIB_SRC_DEVEL=y
+CT_NEWLIB_DEVEL_URL="https://github.com/zephyrproject-rtos/newlib-cygwin.git"
+CT_NEWLIB_DEVEL_REVISION="2893dbb10f1fe3bf431a26f89a22e12a5020dc02"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32s2_zephyr-elf.config:17]:
```suggestion
CT_NEWLIB_SRC_CUSTOM=y
CT_NEWLIB_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/newlib"
```

File: configs/xtensa-espressif_esp32s2_zephyr-elf.config [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+CT_LIBC_NEWLIB_MULTITHREAD=y
+CT_LIBC_NEWLIB_RETARGETABLE_LOCKING=y
+# CT_LIBC_NEWLIB_WIDE_ORIENT is not set
+# CT_LIBC_NEWLIB_NANO_MALLOC is not set
+# CT_LIBC_NEWLIB_NANO_FORMATTED_IO is not set
+CT_LIBC_NEWLIB_EXTRA_SECTIONS=y
+CT_GCC_SRC_DEVEL=y
+CT_GCC_DEVEL_VCS_git=y
+CT_GCC_DEVEL_URL="https://github.com/zephyrproject-rtos/gcc.git"
+CT_GCC_DEVEL_REVISION="8c1f3dcc6eddf3392d77f56c9bed69e5d6446cdf"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32s2_zephyr-elf.config:36]:
```suggestion
CT_GCC_SRC_CUSTOM=y
CT_GCC_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gcc"
```

File: configs/xtensa-espressif_esp32s2_zephyr-elf.config [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+CT_CC_GCC_EXTRA_CONFIG_ARRAY="--with-gnu-ld --with-gnu-as --enable-initfini-array"
+CT_CC_LANG_CXX=y
+CT_DEBUG_GDB=y
+CT_GDB_SRC_DEVEL=y
+CT_GDB_DEVEL_URL="https://github.com/zephyrproject-rtos/binutils-gdb.git"
+CT_GDB_DEVEL_REVISION="76b05e962501d75d995b9f9dd13f837df36b7208"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32s2_zephyr-elf.config:42]:
```suggestion
CT_GDB_SRC_CUSTOM=y
CT_GDB_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gdb"
```

File: configs/xtensa-espressif_esp32s2_zephyr-elf.config [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+CT_GDB_CROSS_PYTHON_BINARY="python3.8"
+CT_GDB_CROSS_BUILD_NO_PYTHON=y
+CT_COMP_LIBS_NEWLIB_NANO=y
+CT_NEWLIB_NANO_SRC_DEVEL=y
+CT_NEWLIB_NANO_DEVEL_URL="https://github.com/zephyrproject-rtos/newlib-cygwin.git"
+CT_NEWLIB_NANO_DEVEL_REVISION="2893dbb10f1fe3bf431a26f89a22e12a5020dc02"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32s2_zephyr-elf.config:48]:
```suggestion
CT_NEWLIB_NANO_SRC_CUSTOM=y
CT_NEWLIB_NANO_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/newlib"
```

**Review by stephanosio — CHANGES_REQUESTED** [review]
There has been a slight change to the way the toolchain components are referenced in the configs (#444).

Could you update the PR as follows?

File: configs/xtensa-espressif_esp32_zephyr-elf.config [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+# CT_PREFIX_DIR_RO is not set
+# CT_LOG_PROGRESS_BAR is not set
+CT_TARGET_CFLAGS="-mlongcalls"
+CT_ARCH_XTENSA=y
+CT_XTENSA_CUSTOM=y
+CT_OVERLAY_NAME="espressif_esp32"
+CT_OVERLAY_LOCATION="./overlays"
+CT_TARGET_VENDOR="espressif_esp32_zephyr"
+CT_GDB_SRC_CUSTOM=y
+CT_GDB_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gdb"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32_zephyr-elf.config:13]: You have specified GDB source configs twice, this needs to be:
```
CT_BINUTILS_SRC_CUSTOM=y
CT_BINUTILS_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/binutils"
```

File: configs/xtensa-espressif_esp32_zephyr-elf.config [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+CT_LIBC_NEWLIB_EXTRA_SECTIONS=y
+CT_GCC_SRC_CUSTOM=y
+CT_GCC_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gcc"
+CT_CC_GCC_EXTRA_CONFIG_ARRAY="--with-gnu-ld --with-gnu-as --enable-initfini-array"
+CT_CC_LANG_CXX=y
+CT_DEBUG_GDB=y
+CT_GDB_SRC_CUSTOM=y
+CT_GDB_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gdb"
+CT_GDB_CROSS_PYTHON_BINARY="python3.8"
+CT_GDB_CROSS_BUILD_NO_PYTHON=y
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32_zephyr-elf.config:39]: Please remove this too as per https://github.com/zephyrproject-rtos/sdk-ng/commit/5abd8f7db6f53d7ba994a93d82d7fde783a3c691.

File: configs/xtensa-espressif_esp32s2_zephyr-elf.config [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+CT_LIBC_NEWLIB_EXTRA_SECTIONS=y
+CT_GCC_SRC_CUSTOM=y
+CT_GCC_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gcc"
+CT_CC_GCC_EXTRA_CONFIG_ARRAY="--with-gnu-ld --with-gnu-as --enable-initfini-array"
+CT_CC_LANG_CXX=y
+CT_DEBUG_GDB=y
+CT_GDB_SRC_CUSTOM=y
+CT_GDB_CUSTOM_LOCATION="${GITHUB_WORKSPACE}/gdb"
+CT_GDB_CROSS_PYTHON_BINARY="python3.8"
+CT_GDB_CROSS_BUILD_NO_PYTHON=y
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32s2_zephyr-elf.config:39]: Please remove this too as per https://github.com/zephyrproject-rtos/sdk-ng/commit/5abd8f7db6f53d7ba994a93d82d7fde783a3c691.

File: configs/xtensa-espressif_esp32_zephyr-elf.config [hunk 14]
```diff
@@ -0,0 +1,66 @@
+CT_CONFIG_VERSION="3"
+CT_EXPERIMENTAL=y
+CT_LOCAL_TARBALLS_DIR="${CT_PREFIX:-${HOME}/x-tools}/sources"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32_zephyr-elf.config:3]: Please remove this as per https://github.com/zephyrproject-rtos/sdk-ng/commit/006aef1658fd51999d482dd28642e9d834adde3a.

File: configs/xtensa-espressif_esp32s2_zephyr-elf.config [hunk 15]
```diff
@@ -0,0 +1,66 @@
+CT_CONFIG_VERSION="3"
+CT_EXPERIMENTAL=y
+CT_LOCAL_TARBALLS_DIR="${CT_PREFIX:-${HOME}/x-tools}/sources"
```
Comment (stephanosio) [inline configs/xtensa-espressif_esp32s2_zephyr-elf.config:3]: Please remove this as per https://github.com/zephyrproject-rtos/sdk-ng/commit/006aef1658fd51999d482dd28642e9d834adde3a.

**Review by stephanosio — CHANGES_REQUESTED** [review]
Thanks for the update. It looks like the GDB source configs were duplicated in the `xtensa-espressif_esp32_zephyr-elf.config` (specified GDB instead of Binutils at the top).

Also I found a few extra changes to be made; it would be good to address them while we are at it.

**Comment by sylvioalves** [comment]
> Thanks for the update. It looks like the GDB source configs were duplicated in the `xtensa-espressif_esp32_zephyr-elf.config` (specified GDB instead of Binutils at the top).
> 
> Also I found a few extra changes to be made; it would be good to address them while we are at it.

Thanks for pointing that out.

**Review by stephanosio — APPROVED** [review]
Looks ok overall.

Pending tests.

**Comment by stephanosio** [comment]
Ran full twister build on the `esp32` and `esp32s2_saola` boards with https://github.com/zephyrproject-rtos/zephyr/pull/43534/commits/6e952a06d0a435ffdfb777a1eb99915f4207dca3 (#43534), verified that all test builds succeed (build only, no hardware runs).

**Pull request merged**